### ISSUE TITLE
The sequence `chain(task.s(), group(chain(task.s())))` does not provide arguments for the second task

### ISSUE THREAD

**Issue by tbazadaykin** [celery/celery#3968]
In the [Documentations](http://docs.celeryproject.org/en/latest/userguide/canvas.html#the-primitives) says:

> Groups and chords accepts partial arguments too, so in a chain the return value of the previous task is forwarded to all tasks in the group...

but it doesn't if group contain chain.

**Steps to reproduce**

```
# ct.py
from __future__ import absolute_import, unicode_literals
from celery import Celery

app = Celery(
    'ct',
    broker='amqp://guest@localhost//',
    backend='cache+memcached://127.0.0.1:11211/'
)

@app.task
def inc(x):
   x += 1
   return x

@app.task
def nothing(x):
   return x

if __name__ == '__main__':
    app.start()
```

Start worker:
`python ct.py worker -l info`

In python console:

```
>>> from ct import inc, nothing
>>> from celery import group, chain
>>> t = chain(inc.s(),group(chain(nothing.s(), inc.s()), inc.s()))
>>> r = t.delay(0)
>>> 
```

In `celery` log:

```
[2017-04-10 12:43:36,126: INFO/MainProcess] Received task: ct.inc[5c1beeff-d785-4d81-af2c-35a5bba839ab]  
[2017-04-10 12:43:37,257: WARNING/PoolWorker-1] /home/vagrant/ct-env/local/lib/python2.7/site-packages/celery/app/trace.py:542: RuntimeWarning: Exception raised outside body: TypeError('nothing() takes exactly 1 argument (0 given)',):
Traceback (most recent call last):
  File "/home/vagrant/ct-env/local/lib/python2.7/site-packages/celery/app/trace.py", line 424, in trace_task
    parent_id=uuid, root_id=root_id,
  File "/home/vagrant/ct-env/local/lib/python2.7/site-packages/celery/canvas.py", line 987, in apply_async
    args=args, kwargs=kwargs, **options))
  File "/home/vagrant/ct-env/local/lib/python2.7/site-packages/celery/canvas.py", line 1063, in _apply_tasks
    **options)
  File "/home/vagrant/ct-env/local/lib/python2.7/site-packages/celery/canvas.py", line 567, in apply_async
    dict(self.options, **options) if options else self.options))
  File "/home/vagrant/ct-env/local/lib/python2.7/site-packages/celery/canvas.py", line 597, in run
    first_task.apply_async(**options)
  File "/home/vagrant/ct-env/local/lib/python2.7/site-packages/celery/canvas.py", line 221, in apply_async
    return _apply(args, kwargs, **options)
  File "/home/vagrant/ct-env/local/lib/python2.7/site-packages/celery/app/task.py", line 518, in apply_async
    check_arguments(*(args or ()), **(kwargs or {}))
TypeError: nothing() takes exactly 1 argument (0 given)

  exc, exc_info.traceback)))
[2017-04-10 12:43:37,262: ERROR/PoolWorker-1] Task ct.inc[5c1beeff-d785-4d81-af2c-35a5bba839ab] raised unexpected: TypeError('nothing() takes exactly 1 argument (0 given)',)
Traceback (most recent call last):
  File "/home/vagrant/ct-env/local/lib/python2.7/site-packages/celery/app/trace.py", line 424, in trace_task
    parent_id=uuid, root_id=root_id,
  File "/home/vagrant/ct-env/local/lib/python2.7/site-packages/celery/canvas.py", line 987, in apply_async
    args=args, kwargs=kwargs, **options))
  File "/home/vagrant/ct-env/local/lib/python2.7/site-packages/celery/canvas.py", line 1063, in _apply_tasks
    **options)
  File "/home/vagrant/ct-env/local/lib/python2.7/site-packages/celery/canvas.py", line 567, in apply_async
    dict(self.options, **options) if options else self.options))
  File "/home/vagrant/ct-env/local/lib/python2.7/site-packages/celery/canvas.py", line 597, in run
    first_task.apply_async(**options)
  File "/home/vagrant/ct-env/local/lib/python2.7/site-packages/celery/canvas.py", line 221, in apply_async
    return _apply(args, kwargs, **options)
  File "/home/vagrant/ct-env/local/lib/python2.7/site-packages/celery/app/task.py", line 518, in apply_async
    check_arguments(*(args or ()), **(kwargs or {}))
TypeError: nothing() takes exactly 1 argument (0 given)
```

**Expecting result**

```
>>> r.get()
[2, 2]
```

**Celery info**

```
celery -A ct report

software -> celery:4.0.2 (latentcall) kombu:4.0.2 py:2.7.9
            billiard:3.5.0.2 py-amqp:2.1.4
platform -> system:Linux arch:64bit, ELF imp:CPython
loader   -> celery.loaders.app.AppLoader
settings -> transport:amqp results:cache+memcached://127.0.0.1:11211/

broker_url: u'amqp://guest:********@localhost:5672//'
result_backend: u'cache+memcached://127.0.0.1:11211/'

```

**Notes**

Work fine with Celery 3.x

**Comment by lpsinger** [comment]
I get a similar error when I attempt to accomplish the same thing with a chord:

#### canvastest.py

```python
#!/usr/bin/env python
from celery import Celery

__all__ = ('app', 'ident')

app = Celery('canvastest.py', backend='redis://')

@app.task
def ident(x):
    return x

if __name__ == '__main__':
    app.start()
```

#### Run the worker

```
$ ./canvastest.py worker
```

#### What works: adding a group as a callback for a chord

```
$ ./canvastest.py shell --python
>>> chord([ident.s(1), ident.s(2)])(group(ident.s(), ident.s())).get(timeout=5)
[[2, 1], [2, 1]]
```

#### What is broken: adding a group that contains a chain as a callback for a chord

```
$ ./canvastest.py shell --python
>>> chord([ident.s(1), ident.s(2)])(group(ident.s() | ident.s(), ident.s())).get(5)
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/celery/result.py", line 630, in get
    on_message=on_message,
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/celery/result.py", line 741, in join_native
    on_message, on_interval):
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/celery/backends/async.py", line 143, in iter_native
    for _ in self._wait_for_pending(result, no_ack=no_ack, **kwargs):
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/celery/backends/async.py", line 260, in _wait_for_pending
    raise TimeoutError('The operation timed out.')
TimeoutError: The operation timed out.
```

Error message in the worker:

```
[2017-07-21 17:42:51,247: ERROR/ForkPoolWorker-3] Chord callback for 'f34ebc60-0a7b-47c3-981f-97418a501168' raised: TypeError('ident() takes exactly 1 argument (0 given)',)
Traceback (most recent call last):
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/celery/backends/redis.py", line 281, in on_chord_part_return
    callback.delay([unpack(tup, decode) for tup in resl])
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/celery/canvas.py", line 182, in delay
    return self.apply_async(partial_args, partial_kwargs)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/celery/canvas.py", line 987, in apply_async
    args=args, kwargs=kwargs, **options))
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/celery/canvas.py", line 1063, in _apply_tasks
    **options)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/celery/canvas.py", line 567, in apply_async
    dict(self.options, **options) if options else self.options))
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/celery/canvas.py", line 597, in run
    first_task.apply_async(**options)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/celery/canvas.py", line 221, in apply_async
    return _apply(args, kwargs, **options)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/celery/app/task.py", line 517, in apply_async
    check_arguments(*(args or ()), **(kwargs or {}))
TypeError: ident() takes exactly 1 argument (0 given)
```

**Comment by lpsinger** [comment]
Still broken in celery 4.1.0.

**Comment by lpsinger** [comment]
Possibly a duplicate of, or at least related to, #3902?

**Comment by lpsinger** [comment]
I think that we can isolate this to a bug in `group` or `chain` alone, and not worry about `chord` for the time being. These work:

```
>>> ident.s().delay(1).get(timeout=5)
1
>>> (ident.s() | ident.s()).delay(1).get(timeout=5)
1
>>> group(ident.s(), ident.s()).delay(1).get(timeout=5)
[1, 1]
```

But this is broken:

```
>>> group(ident.s() | ident.s(), ident.s()).delay(1).get(timeout=5)
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/celery/canvas.py", line 182, in delay
    return self.apply_async(partial_args, partial_kwargs)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/celery/canvas.py", line 986, in apply_async
    args=args, kwargs=kwargs, **options))
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/celery/canvas.py", line 1062, in _apply_tasks
    **options)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/celery/canvas.py", line 566, in apply_async
    dict(self.options, **options) if options else self.options))
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/celery/canvas.py", line 596, in run
    first_task.apply_async(**options)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/celery/canvas.py", line 221, in apply_async
    return _apply(args, kwargs, **options)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/celery/app/task.py", line 518, in apply_async
    check_arguments(*(args or ()), **(kwargs or {}))
TypeError: ident() takes exactly 1 argument (0 given)
```

**Comment by auvipy** [comment]
@lpsinger @tbazadaykin could you please review and check the proposed fix at #4481
